exchanges/cryptodotcom/cryptodotcom_websocket.go

```diff
@@ -0,0 +1,613 @@
+	if err != nil {
+		return err
+	}
+	var resp *RespData
+	err = json.Unmarshal(payload, &resp)
+	if err != nil {
+		return err
+	} else if resp == nil {
+		return errors.New("no valid response from server")
+	}
+	if resp.Code != 0 {
+		mes := fmt.Sprintf("error code: %d Message: %s", resp.Code, resp.Message)
+		if resp.DetailCode != "0" && resp.DetailCode != "" {
+			mes = fmt.Sprintf("%s Detail: %s %s", mes, resp.DetailCode, resp.DetailMessage)
+		}
+		return errors.New(mes)
+	}
+	return nil
+}
+
+// Subscribe sends a websocket subscription to a channel message through the websocket connection handlers.
+func (cr *Cryptodotcom) Subscribe(subscriptions []stream.ChannelSubscription) error {
+	return cr.handleSubscriptions("subscribe", subscriptions)
+}
+
+// Unsubscribe sends a websocket unsubscription to a channel message through the websocket connection handlers.
+func (cr *Cryptodotcom) Unsubscribe(subscriptions []stream.ChannelSubscription) error {
+	return cr.handleSubscriptions("unsubscribe", subscriptions)
+}
+
+// GenerateDefaultSubscriptions Adds default subscriptions to websocket to be handled by ManageSubscriptions()
+func (cr *Cryptodotcom) GenerateDefaultSubscriptions() ([]stream.ChannelSubscription, error) {
+	var subscriptions []stream.ChannelSubscription
+	channels := defaultSubscriptions
+	if cr.Websocket.CanUseAuthenticatedEndpoints() {
+		channels = append(
+			channels,
+
+			// authenticated endpoint subscriptions.
+			userBalanceCnl,
+			userOrderCnl,
+			userTradeCnl,
+		)
+	}
+	for x := range channels {
+		if channels[x] == userBalanceCnl {
+			subscriptions = append(subscriptions, stream.ChannelSubscription{
+				Channel: channels[x],
+			})
+			continue
+		}
+		enabledPairs, err := cr.GetEnabledPairs(asset.Spot)
+		if err != nil {
+			return nil, err
+		}
+		for p := range enabledPairs {
+			switch channels[x] {
+			case instrumentOrderbookCnl,
+				tickerCnl,
+				userOrderCnl,
+				userTradeCnl,
+				tradeCnl:
+				subscriptions = append(subscriptions, stream.ChannelSubscription{
+					Channel:  channels[x],
+					Currency: enabledPairs[p],
+				})
+			case candlestickCnl:
+				subscriptions = append(subscriptions, stream.ChannelSubscription{
+					Channel:  channels[x],
+					Currency: enabledPairs[p],
+					Params: map[string]interface{}{
+						"interval": "5m",
+					},
+				})
+			default:
+				continue
+			}
+		}
+	}
+	return subscriptions, nil
+}
+
+func (cr *Cryptodotcom) handleSubscriptions(operation string, subscriptions []stream.ChannelSubscription) error {
+	subscriptionPayloads, err := cr.generatePayload(operation, subscriptions)
+	if err != nil {
+		return err
+	}
+	for p := range subscriptionPayloads {
+		if subscriptionPayloads[p].Authenticated {
+			err = cr.Websocket.AuthConn.SendJSONMessage(subscriptionPayloads[p])
+		} else {
+			err = cr.Websocket.Conn.SendJSONMessage(subscriptionPayloads[p])
+		}
+		if err != nil {
+			return err
+		}
+		time.Sleep(time.Second)
```

Comment:
You pause everything up >= 20 seconds. Rate limiting is done within SendJSONMesage function. This was hard to find, running the application would have shown there was an issue.